cassandra/src/main/scala/org/apache/spark/sql/cassandra/CassandraXDSourceRelation.scala

```diff
@@ -132,32 +136,69 @@ class CassandraXDSourceRelation(
   }
 
   override def sizeInBytes: Long = {
-    //TODO  Retrieve table size from C* system table from Cassandra 2.1.4
     // If it's not found, use SQLConf default setting
     tableSizeInBytes.getOrElse(sqlContext.conf.defaultSizeInBytes)
   }
 
   implicit val cassandraConnector = connector
   implicit val readconf = readConf
-  private[this] lazy val baseRdd =
+
+  lazy val baseRdd =
     sqlContext.sparkContext.cassandraTable[CassandraSQLRow](tableRef.keyspace, tableRef.table)
 
   def buildScan(): RDD[Row] = baseRdd.asInstanceOf[RDD[Row]]
 
-  override def buildScan(requiredColumns: Array[String], filters: Array[Filter]): RDD[Row] = {
-    buildScan(requiredColumns, filters, Map.empty)
+  override def unhandledFilters(filters: Array[Filter]): Array[Filter] = filterPushdown match {
+    case true => predicatePushDown(filters).handledBySpark.toArray
+    case false => filters
   }
 
-  private def resolveUDFsReferences(strId: String, udfs: Map[String, NativeUDF]): Option[FunctionCallRef] =
-    udfs get(strId) map { udf =>
-      val actualParams = udf.children.collect {
-        case at: AttributeReference if(udfs contains at.toString) => Left(resolveUDFsReferences(at.toString, udfs).get)
-        case at: AttributeReference => Left(ColumnName(at.name))
-        case lit: Literal => Right(lit.toString)
-      }
-      FunctionCallRef(udf.name, actualParams)
+  lazy val additionalRules: Seq[CassandraPredicateRules] = {
+    import CassandraSourceRelation.AdditionalCassandraPushDownRulesParam
+    val sc = sqlContext.sparkContext
+
+    /* So we can set this in testing to different values without
+     making a new context check local property as well */
+    val userClasses: Option[String] =
+      sc.getConf.getOption(AdditionalCassandraPushDownRulesParam.name)
+        .orElse(Option(sc.getLocalProperty(AdditionalCassandraPushDownRulesParam.name)))
+
+    userClasses match {
```

Comment:
Style note: I think it can be written using monadic operations as:

```scala
userClasses.map { classes =>
      classes
        .trim
        .split("""\s*,\s*""")
        .map(ReflectionUtil.findGlobalObject[CassandraPredicateRules])
        .reverse
    } getOrElse AdditionalCassandraPushDownRulesParam.default
```